### PR TITLE
Remove hacks about asm vs constant-flow testing

### LINKED ISSUE #9319: Remove hacks about asm vs constant-flow testing
### Suggested enhancement

- [ ] Change `component_test_memsan_constant_flow` to disable `HAVE_ASM` (which should disable all assembly everywhere: `bignum`, `constant_time.c`, etc. - we might still need to manually disable `AESNI_C`).
- [ ] Add a check in `check_config.h` that complains about `HAVE_ASM` with memsan (see [the check in `aesni.c`](https://github.com/Mbed-TLS/mbedtls/blob/c4dfc084648166a78548029ed423f274b64c08a4/library/aesni.c#L418) for inspiration - see also the existing `MBEDTLS_HAS_MEMSAN` in `check_config.h`).
- [ ] Change `component_test_valgrind_constant_flow` to test both with and without `HAVE_ASM`. (Or if that's considered too much for CI load, only without `HAVE_ASM` - if we wrote assembly by hand, hopefully we knew what we were doing, while with C code it's fundamentally impossible to predict how future compiler will translate it.)
- [ ] Then remove the hack in `library/constant_time_impl.h` so that it just predictably obeys `HAVE_ASM` instead of making fancy and possibly surprising decisions based on how it's being tested.

Once this task is completed, there should be no occurrence of `MBEDTLS_TEST_CONSTANT_FLOW_` under `library` except for `version_features.c`. (We may want to add a check for that, with a reminder of the reason.)

### Justification

Library modules should not modify which implementation they select based on how they're being tested. We need predictability in what exactly it is that we are testing.

### PR DESCRIPTION
**Pull request by eleuzi01**
## Description

Resolves #9319 

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required?
- [x] **development PR** here
- [x] **3.6 backport** done -> #9390 
- [x] **2.28 backport** not required
- [x] **tests** Run a Linux release job: https://mbedtls.trustedfirmware.org/blue/organizations/jenkins/mbedtls-restricted-release-new/detail/mbedtls-restricted-release-new/3/pipeline/77

### CODE REVIEW

**Review by waleed-elmelegy-arm — DISMISSED** [review]
LGTM.

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:2220]: The places where PSA vs not-PSA makes a difference, and the places where asm vs not-asm makes a difference, are separate. So we don't really gain anything by testing all four combinations. Since this component is very slow, we prefer to minimize it. So please remove the `psa_no_asm` combination.

File: include/mbedtls/check_config.h
Comment (gilles-peskine-arm) [inline include/mbedtls/check_config.h:249]: “Memsan” is our own abbreviation. The rest of the world knows it as [MemorySanitizer or MSan](https://github.com/google/sanitizers/wiki/MemorySanitizer). Please use “MSan” or “MemorySanitizer” in the error message.

File: tests/scripts/all.sh
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:2171]: Note: `release` components don't run in the pull request CI. Therefore, when we make a direct change to release components, we need to run the Linux part of the [release job](https://mbedtls.trustedfirmware.org/job/mbedtls-restricted-release-new/) on the PR branch.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
That's too much testing, but other than that, looks good to me.

Note that we need to trigger a manual CI (after the rework).

File: tests/scripts/all.sh
Comment (eleuzi01) [reply tests/scripts/all.sh:2171]: I might not have access because it says 'page not found'

**Review by gilles-peskine-arm — DISMISSED** [review]
Looks good to me. Needs a passing Linux release job (either on OpenCI or on the internal CI) before merging.

**Comment by ronald-cron-arm** [comment]
@waleed-elmelegy-arm please have a look to the last version, thanks.

File: tests/scripts/all.sh
Comment (mpg) [reply tests/scripts/all.sh:2171]: You need to be on the internal network or VPN. If you still can't access it, let me know and I'll start the job.

Comment (mpg) [reply tests/scripts/all.sh:2171]: Ah, sorry, commented too fast: the link was to the Open CI so that's not the issue. You need to log in before you can access that job (or manually start a build for any job, for that matter). You should be able to log in with github and your github group membership should give you appropriate permissions on Open CI. If that's not the case, please shout on slack.

**Review by waleed-elmelegy-arm — DISMISSED** [review]
LGTM

File: tests/scripts/all.sh
Comment (mpg) [reply tests/scripts/all.sh:2171]: Forgot you're on holiday this week. I just started the job, then.

**Comment by eleuzi01** [comment]
Should this be backported to 2.28 @gilles-peskine-arm @mpg? The `constant_time_impl.h` does not exist, so it would be partial backport containing changes around MSan. Not sure if the backport should also include Valgrind changes?

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM after rebase from 98a90cf to 3e81411

**Review by waleed-elmelegy-arm — APPROVED** [review]
LGTM

**Pull request merged**
